src/Search/LexiconfreeLabelsyncBeamSearch/LexiconfreeLabelsyncBeamSearch.cc

```diff
@@ -0,0 +1,594 @@
+ * === LexiconfreeLabelsyncBeamSearch ==
+ * =====================================
+ */
+
+const Core::ParameterInt LexiconfreeLabelsyncBeamSearch::paramMaxBeamSize(
+        "max-beam-size",
+        "Maximum number of hypotheses in the search beam.",
+        1, 1);
+
+const Core::ParameterFloat LexiconfreeLabelsyncBeamSearch::paramScoreThreshold(
+        "score-threshold",
+        "Prune any hypotheses with a score that is at least this much worse than the best hypothesis."
+        "If length normalization is enabled, the score threshold is added to the raw score before normalization."
+        "If not set, no score pruning will be done.",
+        Core::Type<Score>::max, 0);
+
+const Core::ParameterInt LexiconfreeLabelsyncBeamSearch::paramSentenceEndLabelIndex(
+        "sentence-end-index",
+        "Index of the sentence-end label in the lexicon."
+        "Can also be inferred from lexicon if it has a lemma with `special='sentence-end'` or `special='sentence-boundary'`");
+
+const Core::ParameterFloat LexiconfreeLabelsyncBeamSearch::paramLengthNormScale(
+        "length-norm-scale",
+        "Exponent of length for the hypothesis length normalization. Scaled scores are computed as score / length^length_norm_scale.",
+        0.0);
+
+const Core::ParameterFloat LexiconfreeLabelsyncBeamSearch::paramMaxLabelsPerTimestep(
+        "max-labels-per-timestep",
+        "Maximum number of emitted labels per input timestep counted via `addInput`/`addInputs`.",
+        1.0);
+
+const Core::ParameterBool LexiconfreeLabelsyncBeamSearch::paramLogStepwiseStatistics(
+        "log-stepwise-statistics",
+        "Log statistics about the beam at every search step.",
+        false);
+
+LexiconfreeLabelsyncBeamSearch::LexiconfreeLabelsyncBeamSearch(Core::Configuration const& config)
+        : Core::Component(config),
+          SearchAlgorithmV2(config),
+          maxBeamSize_(paramMaxBeamSize(config)),
+          scoreThreshold_(paramScoreThreshold(config)),
+          lengthNormScale_(paramLengthNormScale(config)),
+          maxLabelsPerTimestep_(paramMaxLabelsPerTimestep(config)),
+          sentenceEndLabelIndex_(paramSentenceEndLabelIndex(config)),
+          logStepwiseStatistics_(paramLogStepwiseStatistics(config)),
+          debugChannel_(config, "debug"),
+          labelScorer_(),
+          beam_(),
+          extensions_(),
+          newBeam_(),
+          requests_(),
+          recombinedHypotheses_(),
+          initializationTime_(),
+          featureProcessingTime_(),
+          scoringTime_(),
+          contextExtensionTime_(),
+          numHypsAfterScorePruning_("num-hyps-after-score-pruning"),
+          numHypsAfterBeamPruning_("num-hyps-after-beam-pruning"),
+          currentSearchStep_(0ul),
+          totalTimesteps_(0ul),
+          finishedSegment_(false) {
+    beam_.reserve(maxBeamSize_);
+    newBeam_.reserve(maxBeamSize_ * 2);  // terminated + active
+    recombinedHypotheses_.reserve(maxBeamSize_);
```

Comment:
```suggestion
    recombinedHypotheses_.reserve(maxBeamSize_ * 2);
```
`recombination()` is called before the joint pruning and if no hypotheses are recombined, this vector has the same size as `newBeam_`